Refactoring/accountless client

EvgenKor opened the pull request:
Due to huge code size and needed to support interaction with the system without zkAccount, ancient `ZkBobClient` object was divided into two objects:
- `ZkBobProvider` - to support basic interaction with the relayer, prover, pool contract etc without any account data. This object can be used directly but most preferable to use it within `ZkBobClient` which also support user account attaching
- `ZkBobClient` - a child final class inherited from `ZkBobProvider` to support all of zkAccount functional (making transactions, syncing state, generating zkAddresses, etc). This object is identical to the old one from the interface functionality points of view

To reduce source files size and avoid massive final class paradigm the codebase from the old `ZkBobClient` object was diffused between `ZkBobProvider`, `ZkBobClient`, `ZkBobState` and  `IZkBobService` objects.  The state synchronization code wasm moved to the `ZkBobState` object

Added new entity called Service (interface [IZkBobService](https://github.com/zkBob/zkbob-client-js/blob/fc716e2c6bff70b2e9bce9560251028462004f06/src/services/common.ts#L5-L10)) (to represent external REST subsystem). Currently supported two types of services: [Relayer](https://github.com/zkBob/zkbob-client-js/blob/fc716e2c6bff70b2e9bce9560251028462004f06/src/services/relayer.ts#L95) and [DelegatedProver](https://github.com/zkBob/zkbob-client-js/blob/fc716e2c6bff70b2e9bce9560251028462004f06/src/services/prover.ts#L10). Each of them are ready to multi-services configuration (e.g. to support fallbacking ability) @AllFi I have tested local delegated prover behaviour, but could you please additionally confirm all is fine with the prover interaction?

The new `ZkBobClient` object is similar to the old `ZkBobClient` object except a few functions with minor prototype changes (e.g. wei\gwei convert functions become async, removed `tokenAddress` from the all member functions). The global lifecycle approach was changed:
- first we should initialize `ZkBobClient` instance with the [create](https://github.com/zkBob/zkbob-client-js/blob/fc716e2c6bff70b2e9bce9560251028462004f06/src/client.ts#L121) routine (and pass Pool/Chain config and default pool here). It will be initialized as accountless client. In that state you can request any methods which are not belongs to the account (e.g. [networkName](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client-provider.ts#L182), [shieldedAmountToWei/weiToShieldedAmount](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client-provider.ts#L277-L286), [atomicTxFee](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client-provider.ts#L294), [getLimits](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client-provider.ts#L329) etc). If you invoke any account-related method the error will thrown.
- next to attach a user account we should use method [login](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client.ts#L147) with account configuration (spending key, birth index, prover mode, and pool name)
- in any time (even before user login - to support accountless mode) [we can switch between pools](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client.ts#L164) without logout requirement. Pools are addressing by pool names (aliases) which was be passed during client initialization
- to logout from the currently attached account we should invoke [logout method](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client.ts#L157). After logout the client will become accountless again
- it's do not need to remove `ZkBobClient` instance to attach another account: now it could live during all time of application active

**New functional:**
Added ability to [getting balance](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client.ts#L299) and [redeeming gift-cards](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client.ts#L874)

That PR is mostly architectural. It strongly need to make regress testing due to big code changes @vadimnikonov1 to test that PR locally please use the [associated branch of the console](https://github.com/zkBob/zkbob-console/tree/refactoring/accountless-client)

The additional functions introduced in https://github.com/zkBob/zkbob-client-js/pull/108 were removed. That functional currently implemented inside `ZkBobProvider` class (and can be accessed through the `ZkBobClient` methods.

@maxaleks Please confirm the all you need without zkAccount is accessible through `ZkBobClient` object (all of that located in [ZkBobProvider](https://github.com/zkBob/zkbob-client-js/blob/3ad6c6c0244e6f2fd6752a67e590d359db499206/src/client-provider.ts#L61) which extended by the final ZkBobClient class)

The corresponding changes in the console were also prepared in the [refactoring/accountless-client](https://github.com/zkBob/zkbob-console/tree/refactoring/accountless-client). Please feel free to test it locally. Currently console was tuned to support both sepolia and goerli pools

Linked PR: https://github.com/zkBob/zkbob-console/pull/63

AllFi commented on src/state.ts:
```diff
@@ -1,176 +1,706 @@
 import { InternalError } from './errors';
+import { ZkBobRelayer } from './services/relayer';
+import { CONSTANTS } from './constants';
+import { NetworkBackend } from './networks/network';
```
Unused import

EvgenKor replied:
Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/c6a12bb45e4e4ae0d90a7741e0994b85f152bebd

AllFi commented on src/client-base.ts:
```diff
@@ -0,0 +1,524 @@
+
+// Provides base functionality of the zkBob client
+// without attaching the user account
+export class ZkBobAccountlessClient {
```
I'm not sure, but it seems like another abstraction layer would be helpful in this case. There is a lot of similar code related to selecting a pool, relayer, and so on by name. Perhaps it would make sense for the ZkBobAccountlessClient and ZkBobClient to not concern themselves with the existence of multiple pools. We could create something like a ZkBobClientsManager (the name is arbitrary) that could store all the configurations and logic related to switching accounts and pools.

AllFi replied:
But it would require a lot of refactoring again. If you think it's not worth it, I'm fine with that too.

EvgenKor replied:
Actually I think that currently is the best time for global architectural changes (otherwise it can be never done :) ). Before implementation I thought about architectural solutions and came to the current pattern

Could you please explain your proposal in detail? What are the benefits of the separated manager?

Currently the pool switching functional divided between `ZkBobAccountlessClient` and `ZkBobClient` according with responsibility field: accountless client responsible for all except of state. So it acts in a manner like a manager which you propose. As you can see the `ZkBobClient` just works on a selected pool and doesn't provide methods which requires specifying pool (except inherited from the parent class).  Early I suppose `ZkBobAccountlessClient` can be used standalone but currently it's not so. Maybe it makes sense to move few methods inherited with `public` specifier in the `protected` field.  So from the my point of view the pros of the current approach are following:
- Intuitive responsibility fields (`ZkBobAccountlessClient`: all without account, `ZkBobClient`: account only) and inheritance (client with account `extends` the base client functionality in fact)
- A single final class simplify interaction with the library
- Workflow is the same as a supposed UX
- Ability to use the client in the same manner in account-less mode: we do not need a separated object to request pool limits or relayer fee for example

Let's discuss possible architectural improvements and produce the best solution. I think it's an important thing and as I said before it's better to do it now than never. What about extra coding - you can help me anytime :grin:

AllFi replied:
The main question is whether we really need to support the interface with the ability to pass `poolAlias` in every method. If the answer is yes, then the current approach is good. If it is not then here are my thoughts about what we can change.

I think the cons of the current approach are this:
* `ZkBobAccountlessClient` has mostly map fields, and it can't be confident at any time that everything is ready even for the current pool. Therefore we have existence checks in almost every method.
* `ZkBobAccountlessClient.switchToPool` doesn't check the consistency of configuration. That's why we now do something like ` const pool = this.pool(poolAlias);` in `ZkBobClient.switchToPool`. It looks like some abstraction leak.

I suggest to:
* Move all maps (pools, relayers, etc.) in some PoolConfigs object.

AllFi replied:
The main question is whether we really need to support the interface with the ability to pass `poolAlias` in every method. If the answer is yes, then the current approach is good. If it is not then here are my thoughts about what we can change.

I think the cons of the current approach are this:
* `ZkBobAccountlessClient` has mostly map fields, and it can't be confident at any time that everything is ready even for the current pool. Therefore we have existence checks in almost every method.
* `ZkBobAccountlessClient.switchToPool` doesn't check the consistency of configuration. That's why we now do something like ` const pool = this.pool(poolAlias);` in `ZkBobClient.switchToPool`. It looks like some abstraction leak.

I suggest:
* Move all maps (pools, relayers, etc.) in some PoolConfigs object.
* Store this config object in `ZkBobClient`
* Refactor `ZkBobAccountlessClient` so that it only stores pool, relayer, etc. for one pool. In this case, we can freely use all necessary fields without additional validation in every method.
* Pass the config object and current pool to `ZkBobAccountlessClient.constructor` and `ZkBobAccountlessClient.switchToPool`, and validate the consistency of configuration in these methods.
* Remove the `poolAlias` argument and existence checks from the methods of `ZkBobAccountlessClient`.

AllFi replied:
This way, we don't lose any of the cons that you mentioned, but the code will be simpler, from my perspective.

AllFi replied:
We can probably even move all validation in this config class so we'll have a class with only one responsibility: provide consistent pool configuration.

AllFi replied:
> What about extra coding - you can help me anytime 😁
Sure thing!

EvgenKor replied:
> The main question is whether we really need to support the interface with the ability to pass poolAlias in every method

I suppose the end-user application known nothing about `ZkBobAccountlessClient` and interact with `ZkBobClient` only. And of course the application should never pass poolAlias except `switchPool` method

> ZkBobAccountlessClient has mostly map fields, and it can't be confident at any time that everything is ready even for the current pool

Probably I have added some extra checks. But please check the `ZkBobAccountlessClient.constructor`: it validates the initial config (so we cannot have a pool without existing chain for example) and if it checks it successful - all other methods can be sure about configuration is safe

> I suggest:

Thanks for your suggestion. It looks good. I see a single trouble here. In my implementation most of the map members in `ZkBobAccountlessClient` acts like a local cache to improve pool switching performance and reduce duplication. Let's suppose a hypothetical case we have 10 different token pools on the 5 different chains. With your approach we can create up to 50 `ZkBobAccountlessClient` separated instances (each of them will contain network backend, cached values and other overhead). In my approach it will just a single `ZkBobAccountlessClient` instance with five network backends and common cache. On the other hand in your approach we can remove old `ZkBobAccountlessClient` before the pool switching but in that case the cached values (denominators, poolIds, fees, etc) will be also removed and network backends should be re-initialized from the scratch every time on switching.

> Remove the poolAlias argument and existence checks from the methods of ZkBobAccountlessClient.

I reviewed my code once again and I believe I can remove `poolAlias` from the most of places. All methods with `poolAlias` parameter will acts on the currently selected pool

Also I see some naming ambiguity. I think ZkBobAccountlessClient should be renamed in something like 'ZkBobProvider' or something like that. It will expose methods without any pool aliases (just work with the currently selected pool)

So my compromise way is following:
- rename `ZkBobAccountlessClient` to the `ZkBobProvider`
- move all checks to the `constructor` and `switchPool` method
- remove `poolAlias` arg where possible

Wdyt?

AllFi replied:
> But please check the ZkBobAccountlessClient.constructor: it validates the initial config (so we cannot have a pool without existing chain for example) and if it checks it successful - all other methods can be sure about configuration is safe

It's true before the first call of `ZkBobAccountlessClient.switchToPool`.

AllFi replied:
I think it's a good compromise if we can ensure that `ZkBobProvider` will be consistent at any time in his lifetime.

EvgenKor replied:
I will prepare associated changes in the another branch to lookout them before merging here

EvgenKor replied:
@AllFi Please look at [refactoring/accountless-client-rework](https://github.com/zkBob/zkbob-client-js/tree/refactoring/accountless-client-rework) branch (it's compiled and worked without any changes in console)
I'm not sure about additionally checks in `constructor` and `switchPool` method. From the my point of view all checks already performed. If you have no critical objections on it let's merge it here and continue review.

Will wait for the feedback

AllFi replied:
It looks good to me. Let's merge it.

EvgenKor replied:
It's merged. Also I think there are no reasons to hold the provider instance as a separated `ZkBobClient`'s property because almost all of public provider's method (e.g. `networkName`, `shieldedAmountToWei`, `weiToShieldedAmount`, `poolLimits` etc) are used both in `ZkBobClient` and from the outside. So current inheritance approach reduce unneeded proxy code

AllFi replied:
Agree, it doesn't make sense in the current implementation.

AllFi commented:
By the way, maybe it is time to remove `treeVk` from the worker. It seems that we don't use it anyway, but it still requires some unnecessary management (loading on CDN, etc.) from us.

EvgenKor commented:
> By the way, maybe it is time to remove `treeVk` from the worker. It seems that we don't use it anyway, but it still requires some unnecessary management (loading on CDN, etc.) from us.

Yep, It makes sense. Probably we will need it in the future to support relayerless configuration. But I don't think this future is near :) I'll prepare corresponded commit

EvgenKor commented on src/client.ts:
```diff
@@ -1,78 +1,31 @@
-import { ZkBobState } from './state';
+import { SyncStat, ZkBobState } from './state';
 import { TxType } from './tx';
 import { NetworkBackend } from './networks/network';
```
Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/c6a12bb45e4e4ae0d90a7741e0994b85f152bebd

AllFi commented:
Is it ok that we send `eth_blockNumber` requests to all RPCs regardless of the selected pool?
![image](https://user-images.githubusercontent.com/17165678/228805566-8a8b2151-90a6-4a7f-bfa7-0689cba6282d.png)

EvgenKor commented:
> Is it ok that we send `eth_blockNumber` requests to all RPCs regardless of the selected pool?

No of course. It's need to disable unused network backends. Will prepare changes

EvgenKor commented on src/client-base.ts:
```diff
@@ -0,0 +1,524 @@
+    }
+
+    protected pool(poolAlias: string | undefined = undefined): Pool {
+        const token = this.pools[poolAlias ?? this.curPool];
```
Of course. Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/f6ed3f8aaba61f86a1491998a4b8dff0d13abeb6

EvgenKor commented on src/client-base.ts:
```diff
@@ -0,0 +1,524 @@
+                this.relayerFee[actualPool] = cachedFee;
+            } catch (err) {
+                console.error(`Cannot fetch relayer fee, will using default (${DEFAULT_RELAYER_FEE}): ${err}`);
+                return DEFAULT_RELAYER_FEE;
```
Agreed. Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/f6ed3f8aaba61f86a1491998a4b8dff0d13abeb6

EvgenKor commented on src/client.ts:
```diff
@@ -706,18 +560,18 @@ export class ZkBobClient {
 
       // permittable deposit signature should be calculated for the typed data
-      const value = (amountGwei + feeGwei) * state.denominator;
+      const value = (amountGwei + feeGwei) * denominator;
```
Yes, definitely

https://github.com/zkBob/zkbob-client-js/pull/112/commits/f6ed3f8aaba61f86a1491998a4b8dff0d13abeb6

EvgenKor commented:
> By the way, maybe it is time to remove treeVk from the worker. It seems that we don't use it anyway, but it still requires some unnecessary management (loading on CDN, etc.) from us.
Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/4cfcb52271d2445a0c79488b73481181f591300d

> Is it ok that we send eth_blockNumber requests to all RPCs regardless of the selected pool?
Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/2d019079526f55a3a969d402fc75c7b5bc346e26

AllFi commented on src/client.ts:
```diff
@@ -1383,143 +1179,90 @@ export class ZkBobClient {
+  // | Local and delegated prover support                                       |
+  // ----------------------------------------------------------------------------
+  public async setProverMode(mode: ProverMode) {
+    await super.setProverMode(mode);
```
Let's call it after the if block below. If setProverMode throws an error, we won't start loading params.

EvgenKor replied:
It's a good point. Fixed in https://github.com/zkBob/zkbob-client-js/pull/112/commits/309efa90c9a7cbe56c1e560edc023324e7f0c4fc

AllFi reviewed (approved):
It works properly and looks good to me.
